src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/PersistentServiceProviderTokenStore.java

```diff
@@ -0,0 +1,422 @@
+            save();
+        }
+    }
+
+    @Override
+    public void removeExpiredTokens() throws StoreException {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            if (isTokenExpired(token) && tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    @Override
+    public void removeExpiredSessions() throws StoreException {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            if (isSessionExpired(token.getSession()) && tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    @Override
+    public void removeByConsumer(String consumerKey) {
+        load();
+        boolean needToSave = false;
+        for (ServiceProviderToken token : tokenMap.values()) {
+            Consumer consumer = token.getConsumer();
+            if (consumer != null && Objects.equals(consumerKey, consumer.getKey()) &&
+                    tokenMap.remove(token.getToken()) != null) {
+                needToSave = true;
+            }
+        }
+        if (needToSave) {
+            save();
+        }
+    }
+
+    public synchronized void load() {
+        if (tokenMap != null) {
+            return;
+        }
+
+        XmlFile configFile = getConfigFile();
+        if (configFile.exists()) {
+            try {
+                configFile.unmarshal(this);
+            } catch (IOException e) {
+                log.log(SEVERE, "Failed to load OAuth tokens from disk", e);
+                throw new StoreException("Failed to load OAuth tokens");
+            }
+        }
+        // tokenMap will be unmarshalled as a HashMap if the config file exists, otherwise will be null. Either
+        // way, we convert it to the Jenkins-provided concurrent copy-on-write Map that will be copied on each save and
+        // written to disk, until the next Jenkins restart
+        tokenMap = new CopyOnWriteMap.Hash<>(tokenMap != null ? tokenMap : emptyMap());
+    }
+
+    @Override
+    public synchronized void save() {
+        if (BulkChange.contains(this)) {
+            return;
+        }
+
+        try {
+            getConfigFile().write(this);
+        } catch (IOException e) {
+            log.log(SEVERE, "Failed to persist OAuth tokens to disk", e);
+            throw new StoreException("Failed to persist OAuth tokens");
+        }
+        SaveableListener.fireOnChange(this, getConfigFile());
+    }
+
+    /**
+     * The file where {@link ServiceProviderToken tokens} are saved
+     */
+    @VisibleForTesting
+    protected XmlFile getConfigFile() {
+        // TODO: use a more specific directory than root - https://bulldog.internal.atlassian.com/browse/BBSDEV-21416
```

Comment:
Are you to act on this TODO?

Reply:
Also, can you remove the reference to the internal tracker?

Reply:
Ah, fair. I'll remove the TODO and the reference to the ticket

Reply:
P.S. not changing it in this PR (we may not act on it and close as "won't do" later).